lib/writable.js

```diff
@@ -1,69 +1,256 @@
+
+const DEFAULT_BUFFER_SIZE = 16 * 1024; // 16Kb
+
+const errorSymbol = Symbol('error');
+const doWriteSymbol = Symbol('doWrite');
+const validChunkSymbol = Symbol('validChunk');
+
+class WritableBuffer {
+  constructor(bufferSize) {
+    this.offset = 0;
+    this.queue = [];
+    this.writableLength = 0;
+    this.bufferSize = bufferSize;
+    this.buffer = Buffer.allocUnsafe(bufferSize);
+  }
+
+  writeChunk(chunk) {
+    chunk.copy(this.buffer, this.offset);
+    this.writableLength += chunk.length;
+    this.offset += chunk.length;
+  }
+
+  write(chunk, encoding) {
+    const remainingSize = this.bufferSize - this.offset;
+
+    if (chunk.length <= remainingSize) {
+      this.writeChunk(chunk, encoding);
+    } else {
+      const currentBufferChunk = chunk.slice(0, this.remainingSize);
+      this.writeChunk(currentBufferChunk);
+      this.queue.push(this.buffer);
+      this.buffer = Buffer.allocUnsafe(this.bufferSize);
+      this.offset = 0;
+
+      const nextBufferChunk = chunk.slice(remainingSize);
+
+      if (nextBufferChunk.length < this.buffer.length) {
+        this.writeChunk(nextBufferChunk, encoding);
+      } else {
+        this.queue.push(nextBufferChunk);
+        this.writableLength += nextBufferChunk.length;
+      }
+    }
+  }
+
+  getBuffer() {
+    const bufferData = this.buffer.slice(0, this.offset);
+    this.queue.push(bufferData);
+    const ret = this.queue;
+    this.queue = [];
+    this.offset = 0;
+    this.writableLength = 0;
+    return ret;
+  }
+}
+
+class Writable extends EventEmitter {
+  constructor({
+    write = null,
+    writev = null,
+    final = null,
+    destroy = null,
+    defaultEncoding = 'utf-8',
+    bufferSize = DEFAULT_BUFFER_SIZE
+  }) {
+    super();
+
+    if (typeof write === 'function') this._write = write;
+    if (typeof writev === 'function') this._writev = writev;
+    if (typeof final === 'function') this._final = final;
+    if (typeof destroy === 'function') this._destroy = destroy;
+
+    this.corked = 0;
+
+    this.ended = false;
+    this.ending = false;
+    this.writing = false;
+    this.destroyed = false;
+
+    this.bufferSize = bufferSize;
+    this.defaultEncoding = defaultEncoding;
+    this.buffer = new WritableBuffer(bufferSize);
+    this.callbacks = [];
+  }
+
+  cork() {
+    this.corked++;
+  }
+
+  uncork() {
+    if (this.corked) this.corked--;
+    if (this.corked === 0 && this.writableLength) {
+      this.writing = true;
+      this[doWriteSymbol](this.buffer.getBuffer());
+    }
+  }
+
+  setDefaultEncoding(encoding) {
+    if (typeof encoding === 'string')
+      encoding = encoding.toLowerCase();
+    if (!Buffer.isEncoding(encoding))
+      throw new ERR_UNKNOWN_ENCODING(encoding);
+    this.defaultEncoding = encoding;
+    return this;
+  }
+
+  [errorSymbol](err, cb) {
+    if (typeof cb === 'function') {
+      process.nextTick(cb, err);
+    }
+    this.emit('error', err);
+  }
+
+  pipe() {
+    const err = new ERR_STREAM_CANNOT_PIPE();
```

Comment:
Why isn't this operation supported?

Reply:
We can't pipe writable stream.

Reply:
Yeah, my bad, why do we have this method though?
`pipe` should be defined on a `Readable` so why should it be defined here, as it is clearly not a `Readable`?

Reply:
There is such a method in the node js.

Reply:
According to https://nodejs.org/api/stream.html#stream_class_stream_writable there is no such method, correct me if I am wrong.

Reply:
https://github.com/nodejs/node/blob/master/lib/_stream_writable.js

Also @tshemsedinov implement it.
https://github.com/metarhia/metastreams/blob/master/lib/writable.js

OK. I will not implement this method.